### ISSUE TITLE
da.reshape seems to fail, causing console to hang

### ISSUE THREAD

**Issue by lw394** [dask/dask#1468]
Perhaps I'm using the wrong tool for the job, but I've been trying to use dask to manipulate an entire HDF5 stored dataset out of core, and then overwrite the existing HDF5 dataset (i.e. without the complete dataset ever all being in RAM at the same time). 

I'm creating a dask array from the HDF5 dataset, doing many array manipulations, and then intending to da.store(daskarray,hdf5set,compute=True), doing the compute only at this last step.  I can't get there because a da.reshape fails along the way.  Here are the steps to reproduce:

In [2]: import dask.array as da

In [3]: import h5py

In [4]: ds = h5py.File('testdata.hdf5')['data']['xdata']

In [5]: dda = da.from_array(ds,chunks = 1000)

In [6]: dda.shape
Out[6]: (38897, 768)

In [7]: dda = da.reshape(dda,[38897,768,1,1])

In [8]: dda = da.transpose(dda,[0,1,2,3])

In [10]: datar = da.take(dda,range(0,int(dda.shape[1]/2+1)),axis = 1)

In [11]: datai = da.take(dda,range(int(dda.shape[1]/2+1),dda.shape[1]),axis = 1)

In [13]: datai = da.insert(datai,[0,datai.shape[1]],0,axis=1)

In [15]: dda = datar + 1j*datai

In [17]: dda = da.concatenate((dda.real,da.take(dda.imag,range(1,dda.shape[1]-1), axis = 1)),axis=1)

In [18]: dda.shape
Out[18]: (38897, 768, 1, 1)

#NOTE: COMPUTING THE GRAPH AT THIS POINT SEEMS TO WORK FINE

In [19]: dda.compute()
Out[19]: 
array([[[[  1.94494056e+00]],

        [[  8.85800593e+00]],

        [[ -6.20171956e+00]],

        ..., 
        [[  2.40379211e-01]],
...

In [20]: dda.shape
Out[20]: (38897, 768, 1, 1)

# NOW I JUST TRY TO RESHAPE, REMOVING THE SINGLETON DIMENSIONS, AND IT HANGS

In [21]: dda = da.reshape(dda,[38897,768])

---------------------------------
Is this a dask bug, or am I trying to use it in a way it was never intended.  If the latter, can you point me to library better suited for this task?

**Comment by mrocklin** [comment]
It could be a bug.  The da.reshape function is fairly brittle (there are many cases where reshaping out-of-core arrays will fail).  It was a contention addition, see https://github.com/dask/dask/pull/758.

In the case you're trying, you could also just use slicing to add new axes

    # dda = da.reshape(dda,[38897,768,1,1])
    dda = dda[:, :, None, None]

**Comment by lw394** [comment]
thank you for the quick reply.  I realize the above example seems a little silly, but in general those won't be singleton dimensions, and I'll need a legitimate reshape.  As a temporary workaround, to reset the dask graph, and I create a temporary hdf5 dataset with the interim data, then create a  dask array from that dataset, and continue manipulations with a fresh graph. The reshape works just fine in that case. Unfortunately this requires two copies of the  dataset (with different shapes) on disk.

**Comment by mrocklin** [comment]
Yeah.  That is a reasonable workaround.  You might want to say more about the kind of reshaping you're trying to do.  CCing @shoyer in case he has thoughts on this.

**Comment by lw394** [comment]
A common case would be reshaping an dataset of size [M, N1*N2*N3] to a dask array of size [M, N1, N2, N3], doing some manipulation on the data in this format, and then returning back to the HDF file as a [M, N1*N2*N3] array.  But in general I would want to go from 2D to 4D, or 3D to 4D, manipulate, then back to hdf5 dataset shape.

**Comment by shoyer** [comment]
This is not a bug per-se, but a poorly optimized case for the current implementation.

First a general note: If you plan to chunk along the last axes of your array, there is no way the current API of `da.reshape` can work well for you (prior to adding some sort of `order` parameter).

In this case, I think we can do better. Notice what happens to the chunks in this example:
```
In [6]: import dask.array as da

In [7]: x = da.ones((50, 100, 1, 1), chunks=10)

In [8]: x
Out[8]: dask.array<wrapped..., shape=(50, 100, 1, 1), dtype=float64, chunksize=(10, 10, 1, 1)>

In [9]: x.reshape(50, 100)
Out[9]: dask.array<stack-3..., shape=(50, 100), dtype=float64, chunksize=(1, 10)>
```

The problem is that reshape currently works by completely flattening and then unflattening the array. We could make this more efficient by only doing a *partial flatten* -- only flattening the last axes which are actually changed.

**Comment by lw394** [comment]
Interesting. BTW, I don't think I have any need to chunk along the last axis (unless I'm overlooking something). Inspecting my example more carefully, it looks like chunking is always along the 1st axis only. Here's the output after line [17] above:
`In [18]: dda`
`Out[18]: dask.array <concate..., shape=(38897, 768, 1, 1), dtype=float64, chunksize=(1000, 1, 1, 1)>`

I guess my 1st reshape [7] is ok because it occurs at the start of the graph.  In the second where it hangs [21], you're trying to manage many nested operations through a very inefficient reshape, if I understood you correctly.

**Comment by shoyer** [comment]
Yes. The reshape is currently inefficient, though it doesn't need to be.

I started working on a fix. In the process, it occurred to me that it might be a better idea to simply error in cases where we can't do a reshape efficiently. This would probably be preferable to terrible performance.

**Comment by shoyer** [comment]
To elaborate, the circumstances under which it's possible for `da.reshape` to work well are difficult to state plainly. There are a lot of special cases depending on the shape of the input and output. Ultimately, efficient reshape operations in dask.array are those for which it's possible to reshape each chunk independently.

I believe the reshape operation `(M, N1, N2, N3)` -> `(M, N1*N2*N3)` will work if and only if the last two dimensions (N2 and N3) consistent of a single chunk each. For the converse reshape `(M, N1*N2*N3)` -> `(M, N1, N2, N3)` to work, chunks on the last dimension need to be evenly divisible by `N2 * N3`.

**Comment by lw394** [comment]
Thank you for looking at this so quickly, and I look forward to trying the fix.

Also, something else occurred to me.  If you knew in advance that the HDF5 dataset that you're working on (using dask.array) will never change shape (i.e. any reshapes performed in the graph will have to be undone later before writing back to the dataset), could you efficiently avoid reshaping altogether by reinterpreting intermediate tasks with clever slicing?

**Comment by shoyer** [comment]
If you know that every chunk can be operated on independently, you might simply use `dask.array.map_blocks` instead.

**Comment by jcrist** [comment]
Was this fixed by #1469?

**Comment by shoyer** [comment]
@jcrist  Yes
